maint/tuning/coll/ch4/generic.json

```diff
@@ -0,0 +1,230 @@
+               "comm_size=any":
+               {
+                   "avg_msg_size<12288":
+                   {
+                       "composition=MPIDI_Bcast_intra_composition_alpha":{}
+                   },
+                   "avg_msg_size=any":
+                   {
+                       "composition=MPIDI_Bcast_intra_composition_gamma":{}
+                   }
+               }
+            },
+            "comm_hierarchy=node_roots":
+            {
+                "composition=MPIDI_Bcast_intra_composition_gamma":{}
+            },
+            "comm_hierarchy=node":
+            {
+                "composition=MPIDI_Bcast_intra_composition_gamma":{}
+            },
+            "comm_hierarchy=flat":
+            {
+                "composition=MPIDI_Bcast_intra_composition_gamma":{}
+            }
+        }
+    },
+    "collective=allreduce":
+    {
+        "comm_type=intra":
+        {
+            "comm_size=node_comm_size":
+            {
+               "avg_msg_size<=8192":
+               {
+                   "composition=MPIDI_Allreduce_intra_composition_gamma":{}
+               },
+               "avg_msg_size=any":
+               {
+                   "composition=MPIDI_Allreduce_intra_composition_beta":{}
+               }
+            },
+            "comm_size=any":
+            {
+               "comm_hierarchy=parent":
+               {
+                  "is_commutative=yes":
+                  {
+                      "avg_msg_size=any":
+                      {
+                          "composition=MPIDI_Allreduce_intra_composition_alpha":{}
+                      }
+                  },
+                  "is_commutative=no":
+                  {
+                      "composition=MPIDI_Allreduce_intra_composition_beta":{}
+                   }
+               },
+               "comm_hierarchy=node_roots":
+               {
+                   "composition=MPIDI_Allreduce_intra_composition_beta":{}
+               },
+               "comm_hierarchy=node":
+               {
+                   "composition=MPIDI_Allreduce_intra_composition_beta":{}
+               },
+               "comm_hierarchy=flat":
+               {
+                   "composition=MPIDI_Allreduce_intra_composition_beta":{}
+               }
+            }
+        }
+    },
+    "collective=reduce":
+    {
+        "comm_type=intra":
+        {
+            "comm_hierarchy=parent":
+            {
+               "is_commutative=yes":
+               {
+                   "avg_msg_size<=2048":
+                   {
+                       "composition=MPIDI_Reduce_intra_composition_beta":{}
+                   },
+                   "avg_msg_size=any":
+                   {
+                       "composition=MPIDI_Reduce_intra_composition_alpha":{}
+                   }
+               },
+               "is_commutative=no":
+               {
+                    "composition=MPIDI_Reduce_intra_composition_alpha":{}
```

Comment:
should be `gamma`

Reply:
I am wondering how come `reduce` tests with `non-commutative` ops were not failing because of this. Is there a bug some where else?

Reply:
I'll try out specific tests to see what happened.

Reply:
@jain-surabhi-23 I think I found the issue and it's a rather serious problem.  On communicator creation, we call the device hook before the subcomms are created so that the device sees the parent comm before it sees the subcomms.  This means that, the communicator hierarchy is not setup yet when the device hook is called.  This causes the selection tree pruning logic to think that the communicator is a flat communicator, thus going into a completely different branch of the tree.

I'll try to rework the device hook logic to workaround this.  I fear that the solution is going to be gross and hacky.

Reply:
I believe this was the issue we found a while ago, and @yfguo put in a fix (added a coll_init hook which is separate from the comm_creation_hook I think). We are to complete the fix by moving tree pruning to the coll_init hook which we haven't done yet because of our internal code merge. But yes, this is a known issue.

Reply:
Oh, yes.  I see the new hook.  That's perfect!  Yay, @yfguo!

Reply:
This hook should have been generalized as pre and post hooks instead of having a collective specific hook.  I'll try to do that too.

Reply:
pre and post hooks sound better. We already have a complete fix for this issue (with the new hook) internally. We haven't upstreamed it because the master does not have this PR yet, but the prereq PR by Yanfei has been done in master per our request.

Reply:
Fixed.